### ISSUE TITLE
Problem with repeats when clustering WGS reads

### ISSUE THREAD

**Issue by edgardomortiz** [torognes/vsearch#85]
Hello,
I am clustering NGS reads from a Whole Genome Shotgun project and I am observing very different outputs from vsearch and usearch. The input file is a subsample of the total of reads that contains 8 million reads.
When I run `vsearch -cluster_smallmem input.fasta -id 0.3 -centroids vsearch_output.centroids -usersort -strand both -sizein -sizeout`, vsearch returns a file with 20955 centroids, with many clusters looking like this:
```
>5_1103_2270_15478_1;size=132;
TTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTT
AGGGTTTAGGGTTTAGGGTTT
>5_1102_7325_72279_1;size=112;
GTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTT
TAGGGTTTAGGGTTTAGGGTT
>5_1103_15857_33428_1;size=110;
TTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTA
GGGTTTAGGGTTTAGGGTTTA
>5_1101_16715_40029_1;size=92;
ACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACC
CTAAACCCTAAACCCTAAACC
>5_1101_14897_49572_1;size=84;
CTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTA
AACCCTAAACCCTAAACCCTA
>5_1101_8306_14438_1;size=82;
CCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCT
AAACCCTAAACCCTAAACCCT
>5_1110_5878_83720_1;size=82;
CCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCCTAAACCC
TAAACCCTAAACCCTAAACCC
```
I would expect these 7 centroids to form just one cluster, because I also used `-strand both`. Actually the file is full of this kind of centroids.

When I run the same command on usearch `usearch -cluster_smallmem input.fasta -id 0.3 -centroids usearch_output.centroids -usersort -strand both -sizein -sizeout`, usearch produces 537589 centroids, which look much more as I expected, and the repeats that are similar end up collapsed in a single centroid:
```
>5_1103_2270_15478_1;size=755;
TTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTTAGGGTTT
AGGGTTTAGGGTTTAGGGTTT
>5_1106_20848_8132_1;size=17478;
AATTTACCCAGATTTGAATCTTACGCCTTTATGCCCTCTTATCGATGTAGGCGGCATTATCAAATCTTGAAAATGAATAG
TAGACATATAATCCTTGGCAT
>5_1105_9880_5460_1;size=16024;
ATTCAAATCTGGGTAAATTCAAACAAGCGTAAATCGGGTTGTGTAAGATGTTCTGAGAATAACAACGAGTTTACATAAGG
CTGGTCTAGTCAAGTCAAATT
>5_1101_20013_64046_1;size=20343;
TAAAATATAATAGCGACTTGAAAAGAAACAAATAAACGATGACACCAAAACGTATACGATGTAGGGTAAATGTTTACAAA
TGAAACTAGATTCTTAATTGA
>5_1106_15688_92259_1;size=1847;
GAAACCAATCCGCGGCATTTAGTAGCGGTAAAGTTAGACCAAACCATGAAACCAACATAAACATTATTGCCCGGCGTACG
GGGAAGGACGTCAATAGTCAC
>5_1101_4183_51561_1;size=1375;
GATTGCGTACCCGACGACCAAAATTAGGGTCAACGCTACCTGTAGGAAGTGTCCGCATAAAGTGCACCGCATGGAAATGA
AGACGGCCATTAGCTGTACCA
```
Examining the files more carefully I noticed that vsearch didn't even include most of the centroids found by usearch. Could this be a bug in vsearch? I tried changing the --qmask options, but the results didn't change at all. Am I missing or misusing any option in vsearch?

Thanks!

**Comment by torognes** [comment]
@edgardomortiz, thanks for reporting this important issue.

I think the main problem here is that there needs to be a certain number of UNIQUE k-mers (k=8 by default) shared between the sequences for vsearch to find any similarities. Since the sequences contain just a repeated 7-mer (TTTAGGG or its reverse complement) there are actually no unique k-mers at all in the first 7 sequences and the approach breaks down. I thought usearch used a similar approach, but apparently it handles this situation better than vsearch. Perhaps it does not use only the unique k-mers as described in the paper, but rather all k-mers. We need change the algorithm to fix this problem.

In addition (but not the main problem here), all of the sequences may be entirely masked by the Dust algorithm used in vsearch. Usearch uses a different masking algorithm by default (fastnucleo) that will not mask these sequences. If you run the “-maskfasta" command with the “-hardmask" option and “-qmask dust", you will see that allmost all of the first 7 sequences are masked. Using “-qmask none” whe clustering will solve that problem.

Usearch and vsearch are also quite different when you use a very low value for id such as 0.3. Usearch seems to ignore many real similarities when using low id thresholds (i.e. id < 0.8). You wrote that usearch ended up with 537589 clusters, and even though vsearch did not cluster the repeated sequences it ended up with just 20955 clusters. The correct number of clusters is probably even lower.

I do not know how you are going to use this data, but a temporary solution might be to run maskfasta with qmask dust and hardmask on all the input sequences and then completely remove those sequences that have a very high number of N's.

**Comment by edgardomortiz** [comment]
Thanks for the advice, I will do just that

**Comment by torognes** [comment]
This problem is related to issue #95, marking this one as duplicate.

**Comment by torognes** [comment]
This problem is solved now in version 1.4.7. This version of vsearch will consider all words in the sequences, not just those occurring exactly once.
